File: leetcode.com/python/270_Closest_Binary_Search_Tree_Value.py
# https://leetcode.com/problems/closest-binary-search-tree-value


class TreeNode(object):

    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

    def insert(self, val):
        if val < self.val:
            if self.left is None:
                self.left = TreeNode(val)
            else:
                self.left.insert(val)
        else:
            if self.right is None:
                self.right = TreeNode(val)
            else:
                self.right.insert(val)
        return self


class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        return self.closestValueHelper(root, target, float("inf"))

    def closestValueHelper(self, root, targetVal, currentClosestNodeVal):
        currentNode = root
        while currentNode is not None:
            currentClosestDiff = abs(targetVal - currentClosestNodeVal)
            currentDiff = abs(currentNode.val - targetVal)
            if currentClosestDiff > currentDiff:
                currentClosestNodeVal = currentNode.val
            if targetVal < currentNode.val:
                currentNode = currentNode.left
            elif targetVal > currentNode.val:
                currentNode = currentNode.right
            else:
                break
        return currentClosestNodeVal


test_tree = TreeNode(4).insert(2).insert(5).insert(1).insert(3)
solution = Solution()

closest_val = solution.closestValue(test_tree, 3.714286)
print(str(closest_val) + " is the Closest Value")

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/270_Closest_Binary_Search_Tree_Value.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func closestValue(_ root: TreeNode?, _ target: Double) -> Int {
        guard let root = root else {
            return Int(target)
        }
        
        if let left = root.left, target < Double(root.val) {
            let leftVal = closestValue(left, target)
            if abs(target - Double(leftVal)) < abs(target - Double(root.val)) {
                return leftVal
            } else {
                return root.val
            }
        } else if let right = root.right, target > Double(root.val) {
            let rightVal = closestValue(right, target)
            if abs(target - Double(rightVal)) < abs(target - Double(root.val)) {
                return rightVal
            } else {
                return root.val
            }
        } else {
            return root.val
        }
    }
}
